[storage/adb] add metadata to fixed operation commit

roberto-bayardo opened the pull request:
Operation types that allow for variable-length values support a convenient "metadata" to be provided with a commit.  This PR adds that capability to fixed-length operation types, and then adds `get_metdata` to the Db trait.

The main downside is that for fixed-length value databases with keys less than 9 bytes in length, this adds an extra `9.saturating_sub(len(Key))` bytes per operation. (For example, if your key is a U64, this requires 1 byte wasted per non-commit operation.)

This PR also changes get_metadata behavior to return only the metadata instead of a (metadata, location of last commit) pair. Instead, the `commit` operation of the Db trait is changed to return the start_range of the committed location to facilitate range proof generation and syncing of the new operations.

roberto-bayardo commented on storage/src/adb/current/ordered.rs:
```diff
@@ -422,10 +427,10 @@ impl<
                 (info.loc, info.chunk, element)
             }
-            ExclusionProofInfo::Commit((loc, chunk)) => {
+            ExclusionProofInfo::Commit((loc, metadata, chunk)) => {
```
I don't see anything getting shadowed here.

danlaine commented on examples/sync/src/bin/server.rs:
We can actually avoid cloning the operations altogether by recording the length before passing to `add_oeprations`.

```suggestion
    DB: Syncable,
    E: Storage + Clock + Metrics + RngCore,
{
    let mut last_time = state.last_operation_time.write().await;
    let now = context.current();

    if now.duration_since(*last_time).unwrap_or(Duration::ZERO) >= config.op_interval {
        *last_time = now;

        // Generate new operations
        let new_operations =
            DB::create_test_operations(config.ops_per_interval, context.next_u64());
        let new_operations_len = new_operations.len();

        // Add operations to database and get the new root
        let root = {
            let mut database = state.database.write().await;
            if let Err(err) = DB::add_operations(&mut *database, new_operations).await {
                error!(?err, "failed to add operations to database");
            }
            DB::root(&*database)
        };

        state.ops_counter.inc_by(new_operations_len as u64);
        let root_hex = root
            .as_ref()
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect::<String>();
        info!(
            new_operations_len,
```

danlaine commented on examples/sync/src/bin/server.rs:
```diff
@@ -289,6 +290,7 @@ async fn handle_client<DB, E>(
 ) -> Result<(), Box<dyn std::error::Error>>
 where
     DB: Syncable + Send + Sync + 'static,
+    DB::Operation: Decode + Read<Cfg = ()> + Clone + Send,
```
```suggestion
    DB::Operation:Read<Cfg = ()> + Send,
```

roberto-bayardo replied:
done

danlaine commented on storage/fuzz/fuzz_targets/adb_any_fixed_sync.rs:
```diff
@@ -168,7 +168,7 @@ fn fuzz(input: FuzzInput) {
 
                 Operation::Commit => {
-                    db.commit().await.expect("Commit should not fail");
+                    db.commit(None).await.expect("Commit should not fail");
```
Should we add a fuzz operation to check that the most recently written metadata is correct? We could have the metadata be a counter that tracks the number of commits, or something.

roberto-bayardo replied:
done, though I can't run fuzz tests locally for some reason (getting some verbose linker error?). Let's see if it passes on the CI...

danlaine commented on storage/src/adb/operation/fixed/ordered.rs:
```diff
@@ -22,73 +22,81 @@ pub enum Operation<K: Array + Ord, V: CodecFixed> {
-            "array size too small for commit op"
-        );
+impl<K: Array, V: CodecFixed> Operation<K, V> {
+    // Commit op has a context byte, an option indicator, a metadata value, and a u64 location.
```
Nit: I think this is the more commonly used word? I think it should be clear what this is either way.

```suggestion
    // Commit op has a discriminant, an option indicator, a metadata value, and a u64 location.
```

danlaine commented on storage/src/adb/operation/fixed/unordered.rs:
```diff
@@ -21,42 +21,41 @@ pub enum Operation<K: Array, V: CodecFixed> {
-            "array size too small for commit op"
-        );
+impl<K: Array + Ord, V: CodecFixed> Operation<K, V> {
+    // Commit op has a context byte, an option indicator, a metadata value, and a u64 location.
```
Same nit about context byte vs discriminant

danlaine reviewed (commented):
Would you mind adding tests that ensure `get_metadata` doesn't return stale metedata? i.e. Do a commit with `Some` metadata, then another with `None` metadata or different metadata, and ensure that the newest metadata is returned and not the old one.

roberto-bayardo commented:
> Would you mind adding tests that ensure `get_metadata` doesn't return stale metedata? i.e. Do a commit with `Some` metadata, then another with `None` metadata or different metadata, and ensure that the newest metadata is returned and not the old one.

Added this to the "multiple_commits" test in adb::any::unordered::fixed

roberto-bayardo commented on examples/sync/src/bin/server.rs:
```diff
@@ -396,6 +398,7 @@ async fn run_helper<DB, E>(
 ) -> Result<(), Box<dyn std::error::Error>>
 where
     DB: Syncable + Send + Sync + 'static,
+    DB::Operation: Decode + Read<Cfg = ()> + Clone + Send,
```
done

roberto-bayardo commented on storage/fuzz/fuzz_targets/adb_any_fixed_sync.rs:
```diff
@@ -212,7 +233,7 @@ fn fuzz(input: FuzzInput) {
 
                     db = Any::<_, Key, Value, Sha256, TwoCap>::init(
                         context.clone(),
-                        test_config("src"),
```
This seemed to be a bug in the fuzz test?

roberto-bayardo commented on storage/src/adb/store/mod.rs:
```diff
@@ -184,11 +187,13 @@ pub trait Db<K: Array, V: Codec> {
-    /// this function. Also raises the inactivity floor according to the schedule.
+    /// this function. Also raises the inactivity floor according to the schedule. Returns the
+    /// starting location of the committed range of operations. (The end of the range will be
+    /// `op_count`.)
```
We might consider returning a pair of values representing the range of committed operations, where the second value is always op_count().

danlaine commented on storage/src/adb/current/unordered.rs:
```diff
@@ -141,8 +141,8 @@ impl<
     /// Commit pending operations to the adb::any, ensuring their durability upon return from this
-    /// function.
-    async fn commit_ops(&mut self) -> Result<(), Error> {
+    /// function. Returns the start location of the committed range.
```
We return `()` here?

danlaine commented on storage/src/adb/store/mod.rs:
```diff
@@ -184,11 +187,16 @@ pub trait Db<K: Array, V: Codec> {
+    fn commit(
+        &mut self,
+        metadata: Option<V>,
+    ) -> impl Future<Output = Result<(Location, Location), Error>>;
```
I think it might be a bit more Rusty to return `Range<Location>` here. That's also the type we use in the ADB state sync code to represent the range of live operations to be fetched/applied to the log.
